src/Orleans/Serialization/SerializationManager.cs

```diff
@@ -2280,6 +2280,23 @@ private static void RegisterSerializationProviders(List<TypeInfo> providerTypes)
                         logger.Error(ErrorCode.SerMgr_ErrorLoadingAssemblyTypes, "Failed to create instance of type: " + type.FullName, exception);
                     }
                 });
+
+            // for the outside runtime client, a type may not be registered in the serialization manager before attempting to deserialize
+            // which ends up causing deserialization to fail.  to fix this, we need to register the type if it's supported by any serializer.
+            Action<Assembly> registerAction = 
+                assembly =>
+                {
+                    foreach (var type in assembly.GetTypes().Where(
+                        type => 
+                        (type.IsClass || type.IsValueType) && !type.IsAbstract &&
+                        externalSerializers.Any(ser => ser.IsSupportedType(type))))
+                    {
+                        SerializationManager.Register(type);
+                    }
+                };
+
+            AppDomain.CurrentDomain.AssemblyLoad += (o, e) => registerAction(e.LoadedAssembly);
+            Array.ForEach(AppDomain.CurrentDomain.GetAssemblies(), registerAction);
```

Comment:
In order to resolve the types from the type key string, there are a few annoyances:

- All types in namespaces that start with System are truncated, so to resolve types from the system namespace, you'd have to enumerate through all of the system namespaces in all of the assemblies and see if that type exists
- nested types are delimited using the period character which means that for a type key string like Mine.Orleans.Serialization, it is ambiguous as it is the type Mine+Orleans+Serialization in the global namespace, Orleans+Serialization in the Mine namespace, or Serialization in the Mine.Orleans namespace.  All of these combos would have to be evaluated against all loaded assemblies